cpp/src/reductions/histogram.cu

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stream_compaction/stream_compaction_common.hpp>
```

Comment:
~~Can you please remind me why we need this?~~

Oh, OK. We should probably stop adding new implementations with the legacy map. Can you please use the `cuco::experimental::static_map` instead?

Reply:
Such change will affect both this PR and the existing `distinct` family APIs. Ideally, we should switch to `static_set` and provide more benchmark analysis instead.

So I would defer to moving from the legacy `static_map` here.

Reply:
I see that `distinct` could use `static_set` and correct me if I'm wrong, it seems that the histogram needs payloads for count accumulation.

Reply:
By all means, `hash_map_type` is declared in ``hash_reduce_by_row.cuh` thus we don't need to include this header.

Reply:
Unfortunately when refactoring `hash_reduce_by_row.cuh `, I forgot to redeclare `hash_table_allocator_type`. What do your next suggestion here? Redeclare it in....?

Reply:
> I see that distinct could use static_set and correct me if I'm wrong, it seems that the histogram needs payloads for count accumulation.

Histogram and distinct all need something like reduction-set. Firstly, keys (indices) are inserted into the map/set. Then, all the indices are checked again, performing some reduction ops on an external buffer. `distinct` does atomic `min`/`max`/`count` while `histogram` does atomic `count`.

Reply:
> Unfortunately when refactoring `hash_reduce_by_row.cuh `, I forgot to redeclare `hash_table_allocator_type`. What do your next suggestion here? Redeclare it in....?

Yeah, declaring `hash_table_allocator_type` in the `hash_reduce_by_row.cuh` header. The goal is not to Include a `src` header in `include` files.

Reply:
Or simply just adding one line in the current file
```cpp
using hash_table_allocator_type = rmm::mr::stream_allocator_adaptor<default_allocator<char>>;
```

Reply:
Done.